patterns.md

```diff
@@ -0,0 +1,62 @@
+# Constants in patterns
+
+Constants that pass the [structural equality check](https://github.com/rust-lang/rust/issues/74446) can be used in patterns:
+```rust
+const FOO: i32 = 13;
+
+fn is_foo(x: i32) -> bool {
+  match x {
+    FOO => true,
+    _ => false,
+  }
+}
+```
+However, that check has some loopholes, so e.g. `&T` can be used in a pattern no matter the `T`.
+
+A const-pattern is compiled by calling `PartialEq::eq` to compare the subject of the `match` with the constant,
+except for TODO where the constant is treated as if it was inlined as a pattern (and the usual `match` tree is constructed).
+
+## Soundness concerns
+
+Most of the time there are no extra soundness concerns due to const-patterns; except for surprising behavior nothing can go wrong when the structural equality check gives a wrong answer.
+However, there is one exception: some constants participate in exhaustiveness checking.
+If a pattern is incorrectly considered exhaustive, that leads to a critical soundness bug.
+
+Exhaustiveness checking is done for constants of type TODO, as well as `&[u8]` and `&str` (but no other types with references).
```

Comment:
Exhaustiveness checks are done for constants of all types except references and with references only for `&[u8]` and `&str`. I don't know how to possibly exclude these constants as they also have a literal version `b"foo"` and `"foo"`, which are indistinguishable at the pattern checking stage. Even if we somehow set it up that named constants don't get evaluated until necessary, it feels very fragile to require lazy evaluation for such a semantic information.

Reply:
*All* types? I hope not for raw pointers and function pointers, I cannot imagine that to be soundly possible...

Reply:
If you have a link to the relevant code, that might be good to explain (AFAIK there is no RFC we could link to).

Reply:
There are several points:

* most of the stuff that takes apart constants is https://github.com/rust-lang/rust/blob/2c28244cf0fc9868f55070e55b8f332d196eaf3f/src/librustc_mir_build/hair/pattern/const_to_pat.rs#L204 This is what makes constants indistinguishable from writing the patterns directly
* structural match checks happen here: https://github.com/rust-lang/rust/blob/2c28244cf0fc9868f55070e55b8f332d196eaf3f/src/librustc_mir_build/hair/pattern/const_to_pat.rs#L121
* actual reference comparisons via `PartialEq::eq` happen [here](https://github.com/rust-lang/rust/blob/2c28244cf0fc9868f55070e55b8f332d196eaf3f/src/librustc_mir_build/build/matches/test.rs#L355) but this is a red herring, that's just codegen, even though lots of failures happen there if the previous logic mucks up
* some lazy dereferencing happens [here](https://github.com/rust-lang/rust/blob/2c28244cf0fc9868f55070e55b8f332d196eaf3f/src/librustc_mir_build/hair/pattern/_match.rs#L321) in the middle of match logic. Likely this is also a red herring, but I fixed some ICEs there before fully understanding what I was doing
* checking slice patterns against constants of slice type is done [here](https://github.com/rust-lang/rust/blob/2c28244cf0fc9868f55070e55b8f332d196eaf3f/src/librustc_mir_build/hair/pattern/_match.rs#L2123)

Reply:
Hm, none of that seems to have to do anything with exhaustiveness checking?

Reply:
Well, they all have effects on exhaustiveness (because they change the datastructures that exhaustiveness works on) or are invoked by exhaustiveness. The actual exhaustivness entry points are https://github.com/rust-lang/rust/blob/2c28244cf0fc9868f55070e55b8f332d196eaf3f/src/librustc_mir_build/hair/pattern/_match.rs#L823 and https://github.com/rust-lang/rust/blob/2c28244cf0fc9868f55070e55b8f332d196eaf3f/src/librustc_mir_build/hair/pattern/_match.rs#L974 and https://github.com/rust-lang/rust/blob/2c28244cf0fc9868f55070e55b8f332d196eaf3f/src/librustc_mir_build/hair/pattern/_match.rs#L2069 (the latter being where constants and slices and arrays are hardcoded)

Reply:
Okay somehow I expected a `match` where I'd clearly see references being treated separately. I guess I'll just transcribe what you said into the text, there is no way for me to verify that.

Reply:
I adjusted the text, please check. :)